### PR TITLE
[FLOC-2993] Document containerized installer

### PR DESCRIPTION
**Pull request by lukemarsden**
**Note that as a change to the labs docs this only needs to be checked to see that it doesn't break the Flocker build.**

Recent changes to the unofficial Flocker tools installer mean that:

* The installer itself can be run inside a container
* The installer can use Terraform to automatically provision nodes on AWS
* The installer now works a lot more reliably, in particular on Ubuntu 14.04

This PR documents these changes.

This PR also includes changes as agreed with @alicesowerby and @robhaswell to run an experiment where the main installation pages offer the chance for people to try the Labs installer.

### CODE REVIEW

**Comment by itamarst** [comment]
This includes a bunch of random arbitrary changes that are not just for labs docs.

File: docs/_themes/clusterhq/static/css/docs.css [hunk 1]
```diff
@@ -275,10 +275,6 @@ p {
     display: none;
 }
 
-.admonition a {
```
Comment (adamtheturtle) [inline docs/_themes/clusterhq/static/css/docs.css]: Please separate this and put it against https://clusterhq.atlassian.net/browse/FLOC-1292

File: CONTRIBUTING.rst [hunk 2]
```diff
@@ -135,7 +135,7 @@ If you have any feature requests or suggestions, we would love to hear about the
 At a minimum you can simply submit a GitHub Pull Request with your changes.
 In order to maximize your chances of getting your code accepted, and to keep you from wasting time:
 
-* Discuss your ideas with us in advance by using our `UserVoice`_ forum, or by filing a `GitHub issue`_.
```
Comment (adamtheturtle) [inline CONTRIBUTING.rst]: This doesn't seem relevant

File: docs/_themes/clusterhq/static/css/docs.css (hunk 1 above)
Comment (lukemarsden) [inline docs/_themes/clusterhq/static/css/docs.css]: i've removed it from this PR

**Comment by lukemarsden** [comment]
@itamarst I've removed the unrelated changes. Sorry about that. Everything in this PR should be as per the description now.

File: docs/labs/installer.rst [hunk 3]
```diff
@@ -1,55 +1,87 @@
...
-This will install the following tools on your machine:
+  .. prompt:: bash $
 
-* ``flocker-sample-files``: put some sample ``cluster.yml`` files in the current directory
-* ``flocker-config``: generate certificates and push them
-* ``flocker-install``: install Flocker OS packages on target nodes
-* ``flocker-plugin-install``: install experimental Docker and the :ref:`Flocker Docker plugin <labs-docker-plugin>` on target nodes
-* ``flocker-tutorial``: print out some instructions on how to test the cluster with ``curl`` commands
-* ``flocker-volumes``: an experimental volumes CLI
+      curl -sSL https://get.flocker.io/ | sh
```
Comment (myechuri) [inline docs/labs/installer.rst]: On AWS, default login is ``ubuntu``. i had to toggle between ``ubuntu`` and ``root`` quite a bit yesterday while following unofficial installer documentation. Would be nice to call it out here.

```
$ sudo curl -sSL https://get.flocker.io/ | sudo sh
```

Comment (myechuri) [inline docs/labs/installer.rst]: This command has been hanging for over 10 minutes on Ubuntu 14.04 EC2 instance:

```
ubuntu@ip-172-31-1-54:~$ sudo curl -sSL https://get.flocker.io/ | sudo sh
Installed /usr/local/bin/uft-flocker-ca
Installed /usr/local/bin/uft-flocker-deploy
Installed /usr/local/bin/uft-flocker-config
Installed /usr/local/bin/uft-flocker-install
Installed /usr/local/bin/uft-flocker-plugin-install
Installed /usr/local/bin/uft-flocker-sample-files
Installed /usr/local/bin/uft-flocker-tutorial
Installed /usr/local/bin/uft-flocker-volumes
Installed /usr/local/bin/uft-flocker-get-nodes
Installed /usr/local/bin/uft-flocker-destroy-nodes
Pulling Docker image for Flocker installer...
```

File: docs/labs/installer.rst [hunk 4]
```diff
@@ -57,133 +89,179 @@ Make a local directory for your cluster files
...
-* Use Amazon EC2 if you want to use our EBS backend.
-  **Note that VMs must be deployed in the same AZ.**
-* Use an OpenStack deployment (e.g. Rackspace, private cloud) if you want to try our OpenStack backend.
-  VMs must be deployed in the same region.
+    In the following step, do not use a key (.pem file) which is protected by a passphrase.
+    If necessary, generate and download a new keypair in the EC2 console.
 
-.. warning::
-    Make sure you can log into the nodes as **root** with a private key (e.g. on Ubuntu or CoreOS on AWS, log in as the default user, then run ``sudo mkdir /root/.ssh; sudo cp .ssh/authorized_keys /root/.ssh/authorized_keys``).
+Now paste the following variables into your ``terraform.tfvars`` file::
```
Comment (myechuri) [inline docs/labs/installer.rst]: Did not get a chance to test terraform, but when i ran unofficial flocker installer yesterday, ``cluster.yml`` had ``control_agent`` key while the script expected ``control_service``. Not sure if this problem still exists with terraform?

File: docs/labs/installer.rst (hunk 3 above)
Comment (myechuri) [inline docs/labs/installer.rst]: Nevermind, switched from EC2 Ubuntu 14.04 to Vagrant Ubuntu14.04 - can make progress now.

File: docs/labs/installer.rst (hunk 4 above)
Comment (myechuri) [inline docs/labs/installer.rst]: Nevermind, have terraform up inside vagrant Ubuntu 14.04.

File: docs/labs/installer.rst [hunk 5]
```diff
@@ -57,133 +89,179 @@ Make a local directory for your cluster files
...
+Install and Configure Flocker
+=============================
+
+Run the following command:
 
 .. prompt:: bash $
 
-    mv cluster.yml.ebs.sample cluster.yml
-    vim cluster.yml # customize for your cluster
+    uft-flocker-install cluster.yml && uft-flocker-config cluster.yml && uft-flocker-plugin-install cluster.yml
```
Comment (myechuri) [inline docs/labs/installer.rst]: This step did not work for me:

```
root@node1:~/clusters/test# uft-flocker-install cluster.yml && uft-flocker-config cluster.yml && uft-flocker-plugin-install cluster.yml
Attempting to connect to 52.89.96.32:22...
Connected to 52.89.96.32:22 after 0.00 seconds!
Attempting to connect to ec2-52-88-174-214.us-west-2.compute.amazonaws.com:22...
Connected to ec2-52-88-174-214.us-west-2.compute.amazonaws.com:22 after 0.00 seconds!
main function encountered error
Traceback (most recent call last):
Failure: twisted.internet.defer.FirstError: FirstError[#0, [Failure instance: Traceback (failure with no frames): <type 'exceptions.Exception'>: Process exited with error code 255: Warning: Identity file /host/Users/madhuriyechuri/hybrid-master.pem not accessible: No such file or directory.
Permission denied (publickey).

]]
root@node1:~/clusters/test# ls -la /host/Users/madhuriyechuri/hybrid-master.pem
-r-------- 1 root root 1671 Sep 22 04:34 /host/Users/madhuriyechuri/hybrid-master.pem
root@node1:~/clusters/test# 
```

Triaging..

File: docs/labs/installer.rst (hunk 3 above)
Comment (lukemarsden) [inline docs/labs/installer.rst]: The script has sudo inside it. Also, that would only hang if `docker` wasn't installed and running. I wonder if I can make the doc clearer that you should check that your Docker install is working...

Comment (lukemarsden) [inline docs/labs/installer.rst]: I added 45c31a67549ba860c94d24275fa3af1a01eba595 to ensure that Docker is running before running the CHQ script.

File: docs/labs/installer.rst (hunk 4 above)
Comment (lukemarsden) [inline docs/labs/installer.rst]: Thanks, `cluster.yml` definitely comes out well-formed now.

File: docs/labs/installer.rst (hunk 5 above)
Comment (lukemarsden) [inline docs/labs/installer.rst]: The path with `/host` in it is the path that exists inside the UFT container. That means that *outside* the UFT container `/Users/madhuriyechuri/hybrid-master.pem` needs to exist. However, it looks like you're running this from inside an Ubuntu VM. If that's the case, you need to specify `private_key_path` as the absolute path to the key inside the VM...

FWIW, this is designed to be seamless if you're using Docker Toolbox, which the official Docker instructions now recommend. That uses `docker-machine` to spin up a VirtualBox instance with `boot2docker` installed on it, with host-path sharing set up.

Can you describe whether you've set up any host path sharing for your Vagrant Ubuntu VM?

Comment (lukemarsden) [inline docs/labs/installer.rst]: I added 1ea7547ddc4be82ee4cfa9cdb0e82d5b43b99449 to clarify that this value must be the absolute path to the key on the *machine running the installer*. Hopefully this helps.

**Comment by lukemarsden** [comment]
Merging since Madhuri took a look over this and gave approval to merge in JIRA. Thanks!

As follow-up, I will make a code change in UFT itself to warn users about keys not existing, giving the user-facing not the container-facing path.

**Comment by lukemarsden** [comment]
Docs build is passing, since this is a docs-only change, other failing tests must be unrelated, so merging.

File: docs/_themes/clusterhq/static/css/docs.css (hunk 1 above)
Comment (lukemarsden) [inline docs/_themes/clusterhq/static/css/docs.css]: done in https://github.com/ClusterHQ/flocker/pull/1994

**Pull request merged**